rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+typedef struct SE_PlatformId {
+ void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+typedef struct SE_TimerFns {
+ const size_t struct_size;
+ uint64_t (*nanoseconds)(SE_Timer timer);
+ uint64_t (*microseconds)(SE_Timer timer);
+} SE_Timer;
+
+typedef struct SE_AllocatorStats {
+  const size_t struct_size;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  bool has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  bool has_bytes_reservable_limit;
+  int64_t bytes_reservable_limit;
+
+  int64_t largest_free_block_bytes;
+} SE_AllocatorStats;
+
+typedef enum SE_EventStatus {
+  SE_EVENT_UNKNOWN,
+  SE_EVENT_ERROR,
+  SE_EVENT_PENDING,
+  SE_EVENT_COMPLETE,
+} SE_EventStatus;
+
+typedef struct SE_Options {
+  const size_t struct_size;
+  int ordinal;
+};
+
+typedef struct SE_Device {
+  const size_t struct_size;
+  char* name;
+  size_t name_len;
+
+  // Device vendor can store handle to their device representation
+  // here.
+  void* device_handle;
+
+  // Any kind of data that plugin device might want to store.
+  void* data;
+};
+
+typedef struct SE_StreamExecutor {
+  const size_t struct_size;
+
+  /*** ALLOCATION CALLBACKS ***/
+  // Synchronously allocates size bytes on the underlying platform and returns
+  // a DeviceMemoryBase representing that allocation. In the case of failure,
+  // nullptr is returned.
+  TF_DeviceMemoryBase* (*allocate)(
+      SE_Device* se, uint64_t size, int64_t memory_space);
+
+  // Deallocate the DeviceMemory previously allocated via this interface.
+  // Deallocation of a nullptr-representative value is permitted.
+  void (*deallocate)(
+      SE_Device* se, TF_DeviceMemoryBase* memory);
+
+
+  // Return allocator statistics.
+  bool (*get_allocator_stats)(SE_Device* executor,
+                              SE_AllocatorStats* stats);
+  // Returns the underlying device memory usage information, if it is available.
+  // If it is not available (false is returned), free/total may not be
+  // initialized.
+  bool (*device_memory_usage)(
+      SE_Device* executor, int64_t* free, int64_t* total);
+
+
+  /*** STREAM CALLBACKS ***/
+  // Creates SE_Stream. This call should also Allocate stream
+  // resources on the underlying platform and initializes its
+  // internals.
+  void (*create_stream)(SE_Device* executor, SE_Stream*, TF_Status*);
+
+  // Deletes SE_Stream and deallocates any underlying resources.
+  void (*delete_stream)(SE_Device* executor, SE_Stream stream);
```

Comment:
Could we change the comment and code from Delete to Destroy? So that Create/Destroy are natural pairs, as `new` is to `delete`.

Reply:
Fixed by @yisitu's commit.